adapters/datablocks/datablocks.go

```diff
@@ -0,0 +1,170 @@
+
+	errs := make([]error, 0, len(request.Imp))
+	headers := http.Header{
+		"Content-Type": {"application/json"},
+		"Accept":       {"application/json"},
+	}
+
+	// Pull the host and source ID info from the bidder params.
+	reqImps, err := splitImpressions(request.Imp)
+
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	requests := []*adapters.RequestData{}
+
+	for reqExt, reqImp := range reqImps {
+		request.Imp = reqImp
+		reqJson, err := json.Marshal(request)
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		urlParams := macros.EndpointTemplateParams{Host: reqExt.Host, SourceId: strconv.Itoa(reqExt.SourceId)}
+		url, err := macros.ResolveMacros(a.EndpointTemplate, urlParams)
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		request := adapters.RequestData{
+			Method:  "POST",
+			Uri:     url,
+			Body:    reqJson,
+			Headers: headers}
+
+		requests = append(requests, &request)
+	}
+
+	return requests, errs
+}
+
+/*
+internal original request in OpenRTB, external = result of us having converted it (what comes out of MakeRequests)
+*/
+func (a *DatablocksAdapter) MakeBids(
+	internalRequest *openrtb.BidRequest,
+	externalRequest *adapters.RequestData,
+	response *adapters.ResponseData,
+) (*adapters.BidderResponse, []error) {
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("ERR, response with status %d", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponse()
+	bidResponse.Currency = bidResp.Cur
+
+	for _, seatBid := range bidResp.SeatBid {
+		for i := 0; i < len(seatBid.Bid); i++ {
+			bid := seatBid.Bid[i]
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getMediaType(bid.ImpID, internalRequest.Imp),
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func splitImpressions(imps []openrtb.Imp) (map[openrtb_ext.ExtImpDatablocks][]openrtb.Imp, error) {
+
+	var m = make(map[openrtb_ext.ExtImpDatablocks][]openrtb.Imp)
+
+	for _, imp := range imps {
+		bidderParams, err := getBidderParams(&imp)
+		if err != nil {
+			return nil, err
+		}
+
+		_, ok := m[*bidderParams]
+		if ok {
+			v, _ := m[*bidderParams]
+			m[*bidderParams] = append(v, imp)
+		} else {
+			m[*bidderParams] = []openrtb.Imp{imp}
+		}
+	}
+
+	return m, nil
+}
+
+func getBidderParams(imp *openrtb.Imp) (*openrtb_ext.ExtImpDatablocks, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Missing bidder ext",
+		}
+	}
+	var datablocksExt openrtb_ext.ExtImpDatablocks
+	if err := json.Unmarshal(bidderExt.Bidder, &datablocksExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Missing host or sourceId Param",
```

Comment:
You might want to reword the error as "Cannot resolve host or sourceId Param", as if the bidder parameters is bad json, they might be there but unreadable. And/or include the error message returned by `json.Unmarshal()` for additional debugging.

Also note that just because err was nil, you may still have nil values in `datablocksExt`. The JSON `{}` would pass `json.Unmarshal()` I believe, but just leave all the properties of `datablocksExt` untouched. I am not sure if you have made any assumptions that the properties will be non-nil if it passes here.

Reply:
Thanks for catching that, will implement better handling cases.